test/integration/certificates/admission_approval_test.go

```diff
@@ -0,0 +1,135 @@
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"context"
+	"testing"
+
+	certv1beta1 "k8s.io/api/certificates/v1beta1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+
+	"k8s.io/kubernetes/plugin/pkg/admission/certificates/approval"
+)
+
+// Verifies that the CSR approval admission plugin correctly enforces that a
+// user has permission to approve CSRs for the named signer
+func TestCSRSignerNameApprovalPlugin(t *testing.T) {
+	tests := map[string]struct {
+		allowedSignerName string
+		signerName        string
+		error             string
+	}{
+		"should admit when a user has permission for the exact signerName": {
+			allowedSignerName: "example.com/something",
+			signerName:        "example.com/something",
+		},
+		"should admit when a user has permission for the wildcard-suffixed signerName": {
+			allowedSignerName: "example.com/*",
+			signerName:        "example.com/something",
+		},
+		"should deny if a user does not have permission for the given signerName": {
+			allowedSignerName: "example.com/not-something",
+			signerName:        "example.com/something",
+			error:             `certificatesigningrequests.certificates.k8s.io "csr" is forbidden: user not permitted to approve requests with signerName "example.com/something"`,
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			const username = "test-user"
+			s, closeFn, _, client := runMasterAdmissionSetup(t, approval.NewPlugin())
+			defer closeFn()
+			grantUserPermissionToApproveFor(t, client, username, test.allowedSignerName)
+			csr := createTestingCSR(t, client.CertificatesV1beta1().CertificateSigningRequests(), "csr", test.signerName)
+
+			// testuserClient is a clientset that impersonates the test user, used for
+			// testing RBAC rules
+			testuserClient := clientset.NewForConfigOrDie(&restclient.Config{
+				Host:          s.URL,
+				ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}},
+				Impersonate: restclient.ImpersonationConfig{
+					UserName: username,
+				},
+			})
+			csr.Status.Conditions = append(csr.Status.Conditions, certv1beta1.CertificateSigningRequestCondition{
+				Type:    certv1beta1.CertificateApproved,
+				Reason:  "AutoApproved",
+				Message: "Approved during integration test",
+			})
+			_, err := testuserClient.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+			if err != nil && test.error != err.Error() {
+				t.Errorf("expected error %q but got: %v", test.error, err)
+			}
+			if err == nil && test.error != "" {
+				t.Errorf("expected to get an error %q but got none", test.error)
+			}
+		})
+	}
+}
+
+func grantUserPermissionToApproveFor(t *testing.T, client clientset.Interface, username string, signerNames ...string) {
```

Comment:
This needs to wait for the RBAC permissions to be active (i.e. do SAR checks for the permission and block until they pass).

(same comment for the other similar code)

Reply:
Done